ssl/quic/quic_ackm.c

```diff
@@ -0,0 +1,1997 @@
+                 * The new interval is between intervals without overlapping or
+                 * touching them, so insert between, preserving sort.
+                 */
+                x = OPENSSL_zalloc(sizeof(struct pn_set_item_st));
+                if (x == NULL)
+                    return 0;
+
+                x->range.start = start;
+                x->range.end   = end;
+
+                x->next = z;
+                x->prev = z->prev;
+                if (x->prev != NULL)
+                    x->prev->next = x;
+                z->prev = x;
+                if (s->head == z)
+                    s->head = x;
+
+                ++s->num_ranges;
+            }
+            break;
+        }
+    }
+
+    return 1;
+}
+
+/*
+ * Remove a range from the set. Returns 0 on allocation failure, in which case
+ * the PN set is unchanged. Otherwise, returns 1. Ranges which are not already
+ * in the set can be removed without issue. If a passed range is not in the PN
+ * set at all, this is a no-op and returns 1.
+ */
+static int pn_set_remove(struct pn_set_st *s, const OSSL_QUIC_ACK_RANGE *range)
+{
+    struct pn_set_item_st *z, *zprev, *y;
+    QUIC_PN start = range->start, end = range->end;
+
+    if (!ossl_assert(start <= end))
+        return 0;
+
+    /* Walk backwards since we will most often be removing at the end. */
+    for (z = s->tail; z != NULL; z = zprev) {
+        zprev = z->prev;
+
+        if (z->range.end < start)
+            /* No overlapping ranges can exist beyond this point, so stop. */
+            break;
+
+        if (start <= z->range.start && end >= z->range.end) {
+            /*
+             * The range being removed dwarfs this range, so it should be
+             * removed.
+             */
+            if (z->next != NULL)
+                z->next->prev = z->prev;
+            if (z->prev != NULL)
+                z->prev->next = z->next;
+            if (s->head == z)
+                s->head = z->next;
+            if (s->tail == z)
+                s->tail = z->prev;
+
+            OPENSSL_free(z);
+            --s->num_ranges;
+        } else if (z->range.start >= start) {
+            /*
+             * The range being removed includes start of this range, but does
+             * not cover the entire range (as this would be caught by the case
+             * above). Shorten the range.
+             */
+            assert(end < z->range.end);
+            z->range.start = end + 1;
+        } else if (z->range.end <= end) {
+            /*
+             * The range being removed includes the end of this range, but does
+             * not cover the entire range (as this would be caught by the case
+             * above). Shorten the range. We can also stop iterating.
+             */
+            assert(start > z->range.start);
+            assert(start > 0);
+            z->range.end = start - 1;
+            break;
+        } else if (z->range.start <= start && z->range.end >= end) {
```

Comment:
Clearer would be `z->range.start < start && z->range.end > end` - the case where the starts or ends are equal are covered by the cases above.